README.md
[![forthebadge made-with-python](http://ForTheBadge.com/images/badges/made-with-python.svg)](https://www.python.org/)

[![PyPI status](https://img.shields.io/pypi/status/unbelipy.svg)](https://pypi.python.org/pypi/unbelipy/)
[![PyPI version fury.io](https://badge.fury.io/py/unbelipy.svg)](https://pypi.python.org/pypi/unbelipy/)
[![PyPI downloads](https://img.shields.io/pypi/dm/unbelipy.svg)](https://pypi.python.org/pypi/unbelipy/)
[![PyPI license](https://img.shields.io/pypi/l/unbelipy.svg)](https://pypi.python.org/pypi/unbelipy/)


# unbelipy
Asynchronous wrapper for UnbelievaBoat's API written in python

## Characteristics
- Easy to use
- Full error handling
- Type hinted readable code
- Active maintenance
- Fully Asynchronous

## Project status
Early beta. It's not yet production ready. 
Although most of the functionality is operational, rate limits are still being worked on. 

## Installation

`pip install unbelipy`

## Use:

```python
from unbelipy import UnbeliClient
import asyncio
TOKEN = "Token generated through Unbelievaboat's portal"

client = UnbeliClient(token=TOKEN)

async def main():
    # get guild information
    guild_info = await client.get_guild(guild_id=305129477627969547)
    print(guild_info)
    # get guild leaderboard
    guild_leaderboard = await client.get_leaderboard(guild_id=305129477627969547)
    print(guild_leaderboard)
    # get user balance
    balance = await client.get_balance(guild_id=305129477627969547, member_id=80821761460928512)
    print(balance)
    # put balance (set to x amount)
    balance = await client.set_balance(guild_id=305129477627969547, 
                                       member_id=80821761460928512,
                                       cash=1000,
                                       reason="Showing off put method")
    # patch balance (increment or decrement by x amount)
    balance = await client.edit_balance(guild_id=305129477627969547, 
                                       member_id=80821761460928512,
                                       cash=-500,
                                       reason="Showing off patch method")
    print(balance)

asyncio.run(main())
```

"balance" is a returned Dataclass with balance information containing:
- total: total amount of currency (cash + bank)
- bank: amount in bank
- cash: amount in cash
- user_id: id of the user for which the amount is set
- guild_id: id for the guild the user belongs to
- rank: rank of the user in the guild according to query parameters
- bucket: the bucket that produced this object

"guild_info" is a dataclass with guild info containing:
- id
- name 
- icon
- owner_id  
- member_count  
- symbol (currency)
- bucket: the bucket that produced this object

### UnbeliClient init parameters:
- `token` unbelivaboat's client token.
- `prevent_rate_limits` (bool) if enabled (True, the default) the client will do its best 
  to prevent 429 type errors (rate limits). This will work even on concurrent tasks or loops.
- `retry_rate_limits` (bool) if enabled (True, default is False) the client will retry requests after 
  getting a 429 error. It will sleep through the retry_after time stipulated by UnbelivaBoat's API

#### Note:
It's recommended to use the client with `prevent_rate_limits` set to True with or without `rety_rate_limits`.
Performance is similar either way but running client with only `retry_rate_limits` may result in multiple 429 errors


### UnbeliClient public attributes
- `rate_limits` this class features attributes about the state of each route. They Update after each request. 
  Bucket Attributes. Each of the following contain an async context manager to prevent 429s in case its enabled and 
  contain information about the specific route rate limit headers.
    `buckets` a dictionary with the bucket name as key and its handler as value
  rate limit Methods:
    `rate_limits.currently_limited()` - returns a list containing the bucket name of the buckets that are currently 
    being limited. 
    `rate_limits.any_limited()` - returns a bool indicating if any bucket is currently being limited
    `rate_limits.is_limited(bucket: str)` - returns a bool indicating if the specified bucket is being limited

# Rate limit buckets examples:
- **get_guild** `'GET/guilds/{guild_id}'` 
- **get_leaderboard** `'GET/guilds/{guild_id}/users'`
- **get_balance** `'GET/guilds/{guild_id}/users/:id'`
- **edit_balance** `'PATCH/guilds/{guild_id}/users/:id'`
- **set_balance** `'PUT/guilds/{guild_id}/users/:id'`
- **get_permissions** `'GET/applications/@me/guilds/{guild_id}'`
  
# Know Issues:
- `'-Infinity'` is accepted by the API as a parameter for cash or bank (edit_balance and set_balance),
  but it does not appear to affect the balance. This is caused because the API receives -Infinity as null which is also 
  used when the value didn't change. At the moment there is no word this is going to be fixed.
  
# Credits
- Currently, global rate limit is handled by Martijn Pieters' [aiolimiter](https://github.com/mjpieters/aiolimiter).

CHANGELOG.md
# 1.2.0b
- Bucket routes edited. Mainly user balance get/patch/put and leaderboard
- Fixed bug in `retry_rate_limits`

# 1.1.1b
- Moved the rate limit throttler for buckets from the exit component of the context manager to enter, after acquiring the lock 

# 1.1.0b
- Rewritten rate limit handlers. client.rate_limits now holds an attribute "buckets" which is a dictionary
  that contains the name of the bucket as key and its handler as value. 
  
  Bucket handlers are classes that hold specific bucket rate limit information updated with each request, and a context manager that throttles requests to prevent 429s if that's enabled.
  
  True concurrency is only possible at the moment with different buckets.
- Dataclasses now contain an attribute "bucket" with the name of the bucket that produced the dataclass.
  This is useful to get the bucket handler from `client.rate_limits.buckets`
- Next patch is expected to finally resolve rate_limit functionality

# 1.0.4b
- Fixed bug with `currently_limited.currently_limited()` from issue #6 

# 1.0.1b - 1.0.3b 
- Documentation improvements
- Versioning reformatting to follow semantic conventions.
- Added Balance string representation information.

# 0.0.1b
- Finally in beta
- Improved README
- Adjusted rate limit prevention parameters
- Added licence (MIT)

# 0.0.9a
- specific route rate limits now managed by semaphore set up after first request
- probably the last version of alpha

# 0.0.8a
- client's attribute rate_limit_data is now rate_limits
- Correcly handles global limit by waiting until cleared. Tested on concurrent spammed tasks. for this it uses [aiolimiter](https://github.com/mjpieters/aiolimiter) 
- rate_limits.global_limit is an async context manager to throttle requests to the API if client.prevent_rate_limits is True
- specific route rate limits are also now handled by waiting for response headers and adjusting speed to prevent 429s
- For now aiolimiter is a requiered dependency, might change in the future

# 0.0.7a
- all client api interaction methods are now executed with client._request which handles rate limits
- Rate limits are now prevented only in non concurrent (massive calls with asyncio.gather will still fail).
- added client parameters to decrease possibility of hitting 429s and retrying if it happens.

# 0.0.6a
- Modified `UnbeliClient.rate_limit_data`. 
  It now holds a class which attributes hold rate limit data in a per bucket basis.
- Added this Changelog


unbelipy/api.py
# -*- coding: utf-8 -*-
"""
Module to facilitate integration with UnbelievaBoat's API
"""

import asyncio
from dataclasses import dataclass, field
from inspect import stack
from json import dumps
from typing import Union, Dict, List, Any, Optional

from aiohttp import ClientSession, ClientResponse

from unbelipy.errors import TooManyRequests, NotFound, api_errors, UnknownError
from unbelipy.rate_limits import BucketHandler, ClientRateLimits

UNB_API_VERSION = 'v1'


@dataclass(order=True)
class Balance:
    """
    Dataclass containing the balance status of the specified member
    Attributes:
        total: total amount of currency (cash + bank)
        bank: amount in bank
        cash: amount in cash
        user_id: id of the user for which the amount is set
        guild_id: id for the guild the user belongs to
        rank: rank of the user in the guild according to query parameters
    Comparisons:
        Comparisons are made only with "total"
        Supported operations:
            __eq__ (==)
            __lt__ (<)
            __le__ (<=)
            __gt__ (>)
            __ge__ (>=)
    Notes:
        Amounts in total, bank and cash can be float if set to infinity (positive or negative)
    """
    total: Union[int, float]
    cash: Union[int, float] = field(compare=False)
    bank: Union[int, float] = field(compare=False)
    user_id: int = field(compare=False)
    guild_id: int = field(compare=False)
    bucket: str = field(compare=False)
    rank: int = field(compare=False, default=None)

    def __post_init__(self):
        for attr in ['cash', 'bank', 'total']:
            value = getattr(self, attr)
            try:
                value = int(value)
            except ValueError:
                d = {'Infinity': float('inf'), '-Infinity': float('-inf')}
                value = d[value]
            finally:
                setattr(self, attr, value)

        for attr in ['user_id', 'guild_id']:
            value = getattr(self, attr)
            setattr(self, attr, int(value))

        if self.rank is not None:
            self.rank = int(self.rank)

    def __repr__(self):
        return f"Balance(total={self.total}, cash={self.cash}, bank={self.bank}, rank={self.rank}, " \
               f"user_id={self.user_id}, guild_id={self.guild_id})"


@dataclass
class UnbGuild:
    """
    Dataclass with guild information. 
    Comparisons are made only with member_count
    """
    id: int = field(compare=False)
    name: str = field(compare=False)
    icon: str = field(compare=False)
    owner_id: int = field(compare=False)
    member_count: int
    symbol: str = field(compare=False)
    bucket: str = field(compare=False)
    channels: List[Any] = field(compare=False, default_factory=[])
    roles: List[Any] = field(compare=False, default_factory=[])

    def __post_init__(self):
        self.id = int(self.id)
        self.owner_id = int(self.owner_id)


def _process_bal(bal_dict: dict, guild_id: int, bucket: str):
    bal_dict.update({'guild_id': guild_id, 'bucket': bucket})
    bal_dict.pop('found', None)
    return Balance(**bal_dict)


def _process_lb(users: dict, guild_id: int, bucket: str):
    return [_process_bal(user, guild_id, bucket) for user in users]


def _check_bal_args(cash, bank, reason) -> bool:
    """Checks types and content of arguments for edit and set balance methods"""
    if cash is None and bank is None:
        raise ValueError('An amount or "Infinity" must be specified for either cash or bank')
    for arg in (d := {'cash': cash, 'bank': bank}):
        value = d[arg]
        if value is None:
            continue
        else:
            if (t := type(value)) not in [int, str]:
                raise TypeError(f"{arg} can only be int or str but was {t}")
            elif t is str and value not in ['Infinity', '-Infinity']:
                raise ValueError(f'When {arg} is a String "Infinity" is expected but "{value}" was received')
    if (t := type(reason)) is not str:
        raise TypeError(f'Reason can only be str but was "{t}"')
    return True


class UnbeliClient:
    """
    Client to interact with Unbelievaboat's API
    Parameters:
        token: Token for Unbelivaboat's API.
        prevent_rate_limits: if True client will sleep through rate limits to prevent 429 errors
        retry_rate_limits: if True client will sleep and retry requests after 429 errors
    Public Attributes:
        rate_limits:
            Dictionary containing information on the rate limit status of the client in the API.
                keys:
                    X-RateLimit-Limit: The number of requests that can be made.
                    X-RateLimit-Remaining: The number of remaining requests that can be made.
                    X-RateLimit-Reset: datetime to when the rate limit resets
                    retry_after: seconds to wait before being able to make another request
    Errors:
        All special error classes are found under unbelipy.errors:
            Exceptions for http status codes:
                400: BadRequest,
                401: Unauthorized,
                403: Forbidden,
                404: NotFound
                429: TooManyRequests,
                500: InternalServerError
    """
    _base_url = 'https://unbelievaboat.com/api/' + f'{UNB_API_VERSION}'
    _base_url_len = len(_base_url)
    _member_url = _base_url + '/guilds/{guild_id}/users/{member_id}'

    def __init__(self,
                 token: str,
                 prevent_rate_limits: bool = True,
                 retry_rate_limits: bool = False):
        self._header = {'Authorization': token}
        self._prevent_rate_limits = prevent_rate_limits
        self._retry_rate_limits = retry_rate_limits
        self.rate_limits = ClientRateLimits(prevent_rate_limits=prevent_rate_limits)

    def _get_member_url(self, guild_id: int, member_id: int):
        url = self._member_url.format(guild_id=guild_id, member_id=member_id)
        route = url[self._base_url_len:-len(str(member_id))] + ':id'
        return url, route

    @staticmethod
    def _get_caller():
        return stack()[2][3]

    def _get_bucket_handler(self, bucket: str):
        bucket_handler = self.rate_limits.buckets.get(bucket)
        if bucket_handler is None:
            bucket_handler = self.rate_limits.buckets[bucket] = BucketHandler(bucket=bucket)
        return bucket_handler

    async def _request(self,
                       method: str,
                       url: str,
                       bucket: str,
                       data: Optional[str] = None,
                       caller=None,
                       **kwargs
                       ) -> Union[int,
                                  Balance,
                                  UnbGuild,
                                  Dict[str, Union[List[Balance], str]],
                                  List[Balance]]:
        """
        Processes requests to the Unbelivaboat's API
        Args:
            method (str): 'PUT', 'PATCH' or 'GET'
            url (str): request url
            data (json str): server data
            kwargs:
                guild_id: id (int) of the guild
                page: int number of page in the case of get_leaderboard
        """
        cs = ClientSession(headers=self._header)
        if caller is None:
            caller = self._get_caller()
        guild_id = kwargs.pop('guild_id', None)
        page = kwargs.pop('page', None)

        if method == 'PUT':
            request_manager = cs.put(url=url, data=data)
        elif method == 'PATCH':
            request_manager = cs.patch(url=url, data=data)
        else:  # defaults to 'GET'
            request_manager = cs.get(url=url)

        bucket_handler: BucketHandler = self._get_bucket_handler(bucket)
        bucket_handler.prevent_429 = self._prevent_rate_limits
        async with self.rate_limits.global_limiter:
            async with bucket_handler as bh:
                async with cs:
                    r = await request_manager
                    bh.check_limit_headers(r)  # sets up the bucket rate limit attributes w/ response headers
                    response_data = await r.json()
                try:
                    if await self._check_response(response=r, bucket=bucket):
                        if caller in ['edit_balance', 'set_balance', 'get_balance']:
                            return _process_bal(response_data, guild_id, bucket)
                        elif caller == 'get_leaderboard':
                            if page is None:
                                return _process_lb(response_data, guild_id, bucket)
                            else:
                                response_data['users'] = _process_lb(response_data['users'], guild_id, bucket)
                                return response_data
                        elif caller == 'get_permissions':
                            return response_data['permissions']
                        elif caller == 'get_guild':
                            response_data['bucket'] = bucket
                            return UnbGuild(**response_data)
                except TooManyRequests as E:
                    if self._retry_rate_limits is True:
                        timeout = response_data['retry_after'] / 1000 + 1
                        await asyncio.sleep(timeout)
                        # reschedule same request
                        return await self._request(method, url, bucket, data, caller=caller, **kwargs)
                    else:
                        raise E

    async def _check_response(self, response: ClientResponse, bucket: str):
        """Checks API response for errors. Returns True only on status 200"""
        status = response.status
        reason = response.reason
        data = await response.json()

        if status == 200:
            return True
        elif status == 429:
            message = data['message']
            if 'global' in data:
                text = f"Global rate limit. {data['message']}"
            elif retry_after := data.get('retry_after'):
                retry_after = int(retry_after) / 1000
                bucket_handler = self._get_bucket_handler(bucket)
                bucket_handler.retry_after = retry_after
                text = f"{message} retry after: {retry_after}s"
            else:
                text = f"{message}"
            raise TooManyRequests(text + f', bucket: {bucket}')
        elif status == 404:
            raise NotFound(f'Error Code: "{status}" Reason: "{reason}", bucket {bucket}')
        else:
            error_text = f'Error code: "{status}" Reason: "{reason}"'
            if status in api_errors:
                raise api_errors[status](error_text)
            else:
                raise UnknownError(error_text)

    async def get_permissions(self, guild_id: int) -> int:
        """
        Returns the application's permissions for the guild with specified id
        """
        if (t := type(guild_id)) is not int:
            raise TypeError(f"guild_id must be an int but {t} was received")
        method = 'GET'
        url = self._base_url + f"/applications/@me/guilds/{guild_id}"
        bucket = method + url[self._base_url_len:]
        return await self._request(method=method, url=url, bucket=bucket)

    async def get_guild(self, guild_id) -> UnbGuild:
        """
        Args:
            guild_id: id of the target guild

        Returns:
            Returns a dataclass for the target guild with the following attributes:
                id (str)
                name (str)
                icon (str: icon hash)
                owner_id (str)
                member_count (int)
                symbol (str) currency symbol
        """
        if (t := type(guild_id)) is not int:
            raise TypeError(f"guild_id must be an int but {t} was received")
        method = 'GET'
        url = f"{self._base_url}/guilds/{guild_id}"
        bucket = method + url[self._base_url_len:]
        return await self._request(method=method, url=url, bucket=bucket)

    async def get_leaderboard(self,
                              guild_id: int,
                              sort: Optional[str] = None,
                              limit: Optional[int] = None,
                              offset: Optional[int] = 1,
                              page: Optional[int] = None
                              ) -> Union[List[Balance],
                                         Dict[str, Union[int,
                                                         List[Balance]]]]:
        """
        Retrieves leaderboard information for a guild
        Args:
            guild_id (int): id of the guild
            sort (str): pass 'cash', 'bank' or 'total' to sort accordingly
            limit (int): amount of users to retrieve
            offset (int): retrieve users only from set place and below in the leaderboard
            page (int): page number to retrieve
                if specified returns a dictionary containing the list of user's leaderboard under key 'users' and
                additional 'page' with the current page and 'total_pages' with number available pages.
        Returns:
            dictionary containing leaderboard information
        """
        bucket_path = f'/guilds/{guild_id}/users'
        url = self._base_url + bucket_path + '?'
        if sort is not None:
            if (t := type(sort)) is not str:
                raise TypeError(f'sort must be type str but was "{t}"')
            elif sort not in ['cash', 'bank', 'total']:
                raise ValueError(f'Sort can only be "cash", "bank" or "total" but was "{sort}"')
            else:
                url += f"sort={sort}&"

        for arg in (d := {'limit': limit, 'offset': offset, 'page': page}):
            if d[arg] is not None:
                if (t := type(d[arg])) is not int:
                    raise TypeError(f'{arg} can only be type int but was {t}')
                else:
                    url += f"{arg}={d[arg]}&"

        method = 'GET'
        url = url[:-1]
        bucket = method + bucket_path
        return await self._request(method=method, url=url, bucket=bucket, guild_id=guild_id, page=page)

    async def get_balance(self, guild_id: int, member_id: int) -> Balance:
        """
        Args:
            guild_id: id of the guild the member belongs to
            member_id: id of the member
        Returns:
            balance (Balance) A dataclass containing the member's balance
        """
        for arg in (d := {'guild_id': guild_id, 'member_id': member_id}):
            if (t := type(d[arg])) is not int:
                raise TypeError(f"{arg} can only be int but was {t}")
        method = 'GET'
        url, route = self._get_member_url(guild_id, member_id)
        bucket = method + route
        return await self._request(method=method, url=url, bucket=bucket, guild_id=guild_id)

    async def edit_balance(self,
                           guild_id: int,
                           member_id: int,
                           cash: Union[int, str] = 0,
                           bank: Union[int, str] = 0,
                           reason: str = '') -> Balance:
        """
        Increase or decrease the Member's balance by a value given in the params.
        To decrease the balance, provide a negative number.

        Args:
            guild_id: id of the guild the member belongs to
            member_id: id of the member
            cash: Amount to modify the member's cash amount
            bank: Amount to modify the member's bank amount
            reason: specifies the reason why the balance was modified

        Returns:
            balance (Balance) a dataclass containing the information of the newly modified balance for the user.


        """

        if _check_bal_args(cash, bank, reason):
            data = {'cash': cash, 'bank': bank, 'reason': reason}

            method = 'PATCH'
            url, route = self._get_member_url(guild_id, member_id)
            bucket = method + route
            return await self._request(method=method, url=url, bucket=bucket, data=dumps(data), guild_id=guild_id)

    async def set_balance(self,
                          guild_id: int,
                          member_id: int,
                          cash: Union[int, str, None] = None,
                          bank: Union[int, str, None] = None,
                          reason: str = '') -> Balance:
        """
        Set the Member's balance to the given params.
        Args:
            guild_id: id of the guild the member belongs to
            member_id: id of the member
            cash: sets the member's cash amount
            bank: sets the member's bank amount
            reason: specifies the reason why the balance was modified
        Usage Notes:
            if cash and/or bank is a string they must be "Infinite".
            At least cash or bank are needed
        Returns:
            balance (Balance) a dataclass containing the information of the newly set balance for the user.
        """

        if _check_bal_args(cash, bank, reason):
            data = {'cash': cash, 'bank': bank, 'reason': reason}
            method = 'PUT'
            url, route = self._get_member_url(guild_id, member_id)
            bucket = method + route
            return await self._request(method=method, url=url, bucket=bucket, data=dumps(data), guild_id=guild_id)
